File: design/kotlin/housepour/HPModal.kt
package ca.lucianlabs.housepour

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.systemBarsPadding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.unit.dp

/** A card centred over the `scrim`, cast shadow deepened ×1.5, fading in `Motion.toast`. Never dark. */
@Composable
fun HPModal(
    isPresented: Boolean,
    onDismiss: () -> Unit,
    content: @Composable ColumnScope.() -> Unit,
) {
    AnimatedVisibility(
        visible = isPresented,
        enter = fadeIn(tween(HPTokens.Motion.TOAST_MS)),
        exit = fadeOut(tween(HPTokens.Motion.TOAST_MS)),
    ) {
        val shape = RoundedCornerShape(HPTokens.Radius.card)
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(HPTokens.Colors.scrim)
                .clickable(interactionSource = remember { MutableInteractionSource() }, indication = null, onClick = onDismiss)
                .systemBarsPadding()
                .imePadding(),
            contentAlignment = Alignment.Center,
        ) {
            Column(
                modifier = Modifier
                    .padding(HPTokens.Space.columnSide)
                    .widthIn(max = HPTokens.Space.columnMax)
                    .fillMaxWidth()
                    .hpShadow(HPTokens.Radius.card, HPTokens.Shadow.contact)
                    .hpShadow(HPTokens.Radius.card, HPTokens.Shadow.cast, alphaScale = 1.5f)
                    .clip(shape)
                    .background(HPTokens.Colors.panel, shape)
                    .border(HPTokens.BORDER_WIDTH.dp, HPTokens.Colors.line, shape)
                    .pointerInput(Unit) { detectTapGestures { } }
                    .verticalScroll(rememberScrollState())
                    .padding(HPTokens.Space.cardPad),
                content = content,
            )
        }
    }
}

File: design/swift/HousePour/HPModal.swift
// House Pour kit — HPModal (COMPONENTS.md).

import SwiftUI

/// A card centred over the scrim, cast shadow deepened. Fades in; never dark.
public struct HPModal<ModalContent: View>: ViewModifier {
    @Binding var isPresented: Bool
    let modal: ModalContent
    public init(isPresented: Binding<Bool>, @ViewBuilder content: () -> ModalContent) {
        _isPresented = isPresented; modal = content()
    }
    public func body(content: Content) -> some View {
        ZStack {
            content
            if isPresented {
                HPTokens.Colors.scrim
                    .ignoresSafeArea()
                    .onTapGesture { isPresented = false }
                    .accessibilityLabel("Dismiss")
                    .accessibilityAddTraits(.isButton)
                let shape = RoundedRectangle(cornerRadius: HPTokens.Radius.card, style: .continuous)
                let card = VStack(alignment: .leading, spacing: 0) { modal }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(HPTokens.Space.cardPad)
                // The card is bounded by the space it is given and scrolls inside when its content
                // is taller than that — the same bound the other two kits carry (`HPModal.kt`'s
                // `verticalScroll`, `.modal-card`'s `max-height` + `overflow-y`). Without it a modal
                // taller than the phone puts its own buttons off-window: the report confirm
                // (PRODUCT §2.15) is 755pt at default Dynamic Type and there is no other route to
                // report anything, so `Send Report` and `Cancel` have to be reachable on a 375×667
                // phone (PRODUCT §5: "iPhone and iPad") and at every type size above it.
                //
                // `ViewThatFits` keeps the natural, content-sized card whenever it fits, so the
                // short confirms are centred and unscrollable exactly as before; only one that
                // would run off the screen becomes a scroller.
                ViewThatFits(in: .vertical) {
                    card
                    ScrollView(.vertical) { card }
                }
                .background(shape.fill(HPTokens.Colors.panel))
                .hpBorder(shape)
                .hpCardShadow(shape: shape, fill: HPTokens.Colors.panel, castMultiplier: HPAlpha.modalCast)
                .frame(maxWidth: HPTokens.Space.columnMax - HPTokens.Space.columnSide * 2)
                .padding(.horizontal, HPTokens.Space.columnSide)
                // The same `columnSide` inset the sides already had, so a full-height card stops
                // short of the safe area rather than running into it.
                .padding(.vertical, HPTokens.Space.columnSide)
                .transition(.opacity)
            }
        }
        .animation(HPMotion.toast, value: isPresented)
    }
}

public extension View {
    func hpModal<C: View>(isPresented: Binding<Bool>, @ViewBuilder content: () -> C) -> some View {
        modifier(HPModal(isPresented: isPresented, content: content))
    }
}
